README.md
# PyPBE

![PyePBE](https://github.com/drericstrong/pypbe/blob/master/images/pypbe_small.jpg)

[![PyPI version](https://badge.fury.io/py/pypbe.svg)](https://badge.fury.io/py/pypbe)
[![Documentation Status](https://readthedocs.org/projects/pypbe/badge/?version=latest)](http://pypbe.readthedocs.io/en/latest/?badge=latest)

PyPBE is a resource for tabletop gaming which allows Gamemasters (GM) to fairly select which random rolling method is closest to an equivalent Point Buy value. Ideally, all players will determine their character's stats exactly the same way. However, in some cases, players may ask for several different options for generating their character's stats. PyPBE is designed to allow GMs to make this decision in a fair way. By calculating the equivalent Point Buy for a random rolling method, the GM can determine the expected power level of the characters, as well.

## Why Would I Use It?

*Sarah, John, and Kara sit down to play the first session of a new Pathfinder campaign with their GM, Lee. Sarah is excited about playing her favorite class, a Ranger. It doesn't matter what stats she rolls- she'll make it work. John is a bit of a min-maxer and wants to roll up a Wizard, so he's keen on getting the highest intelligence possible. Kara wants to play a Monk, which requires decent stats in a few different attributes.* 

*Lee decides that he wants the players to randomly roll for their character's stats. Sarah is excited about the chance for high stats. She's feeling especially lucky today. John, however, is angry. What if he doesn't roll any 18s for his intelligence? Kara is worried. She really wants to play a Monk, but she wants the character to be effective, and bad rolls could significantly reduce the fun she'll have playing the character.*

*What should Lee do?*

Some GMs prefer to let their players choose between rolling for their ability scores and letting them use the Point Buy system. However, not all random rolling methods are created equal. Some (4d6, drop lowest) clearly give higher average results than others (3d6). PyPBE is designed to calculate and visualize the distribution of a specified ability score rolling method, which may provide useful information for decision-making.

The stats that PyPBE calculates aren't the "raw" values of the roll (e.g. typically 3 through 18), they're the "Point Buy Equivalent" of multiple rolls using that rolling method. For instance, if you roll 3d6 six times, you might get 10, 12, 8, 13, 7, 9, which has a Point Buy Equivalent of -2 (0+2-2+3-4-1) using the Pathfinder point buy scheme. 

PyPBE uses Monte Carlo simulation to obtain its results. If you perform the above process thousands/millions of times, you will get a distribution. The mean of that distribution is the fair Point Buy you should select for that rolling method, and 90% of the time, the random roll PBE will fall between the 5%/95% values. The "Typical Array" gives the most likely stat array using that random rolling method.

## Systems
PyPBE is designed for Pathfinder, 3e, 3.5e, 4e, and 5e characters. However, it allows the option to supply a custom Point Buy mapping, which means that it is applicable for any system in which the "Point Buy" concept applies. PyPBE also supports any number of attributes, although it was designed for the common 6-attribute system (strength, constitution, dexterity, intelligence, wisdom, charisma). As the number of attributes increase, note that the plots may become cluttered.

## Try It Out!
A web-based simulator is running [here](https://pypbe.herokuapp.com/pypbe-bk).

![PyePBE](https://github.com/drericstrong/pypbe/blob/master/images/pypbe-bk_screenshot.png)

## Getting Started with the Python API
Current documentation can be found [here](https://pypbe.readthedocs.io/en/latest/).

When using PyPBE as a Python module, the following dependencies are required:

**numpy, seaborn, matplotlib**

PyPBE can be installed using pip:

**pip install pypbe**

Import the PBE class into your program, initializing it using the number and type of dice to roll, then use the "roll_mc" method to investigate the distribution. The results can be visualized using the "plot_histogram" method, and a data row summary can be generated using the "get_results" method. 

For example, running a simulation for three 6-sided dice (3d6), rolled for 6 attributes, can be accomplished using the following code:

> from pypbe import PBE

> alg = PBE(3,6)

> alg.roll_mc()

> alg.plot_histogram()

> results = alg.get_results()

The *plot_histogram()* function will generate a plot that looks like this:

![Example Histogram](https://github.com/drericstrong/pypbe/blob/master/images/4d6_example.png)

The top part of the plot shows the distribution of each dice roll, ranked from lowest (Roll 1) to highest (usually Roll 6). Think about it this way- the mean value of 6 sets of 3d6, repeated many times, will tend towards 10.5 (each six-sided dice has a mean of 3.5, so 3.5 times 3 equals 10.5). However, if you order the 6 sets from lowest to highest, you'll notice that the lowest roll tends to be lower than 10.5, and the highest roll tends to be higher than 10.5. The plot shows the expected value for each ranked roll, which can be interpreted as the "typical" stat array for this rolling method. In the figure above, the 5th and 95th percentiles are given in brackets. For instance, [5,11] means that 90% of the distribution is between 5 and 11.

The distribution of the Point Buy value is shown in the bottom plot by mapping the results of each dice roll to a Point Buy value. The default mapping is: {3:-16, 4:-12, 5:-9, 6:-6, 7:-4, 8:-2, 9:-1, 10:0, 11:1, 12:2, 13:3, 14:5, 15:7, 16:10, 17:13, 18:17}. The mean, standard deviation, 5th, and 95th percentiles are shown on the figure. You can interpret the mean of the Point Buy distribution as the "Point Buy Equivalent"- the Point Buy value that is most fair to choose as the equivalent for the ability score rolling method.

The *get_results()* and *plot_histogram()* methods can be chained with the *roll_mc()* method, like this:

> hist_plot = alg.roll_mc().plot_histogram()

> res = alg.roll_mc().get_results()

## Custom Parameters
More complicated scenarios can be run by adjusting the following user-specified parameters:

### -System (*pbe_map* parameter)
Each RPG system has its own point buy scheme, and buying an attribute value may cost different amounts in different systems. For example, in PF a '10' will cost '0' point buy, while in 3e a '10' will cost '2' point buy. It's important that you specify the right system; the results will vary heavily depending on which system you choose. You can try this out for yourself by comparing PF to 3e. Currently recognized values for this parameter include: 'PF', '3e', '4e', and '5e'. However, a custom point buy scheme can be specified using the *custom_pbe_map* mentioned below. 

**Default**: 'PF'. 

**Example**: the rpg group is playing D&D 3rd edition, so the value should be '3e'.

### -Number of Dice Per Attribute (*num_dice* parameter)
Each time you roll for an attribute (such as STR), this parameter specifies the total number of dice that you will roll. When using 'XdY+Z' notation, this value is 'X'. 

**Example**: if you are rolling three six-sided dice (3d6) to determine your stats, this value should be '3'.

### -Dice Sides (*dice_type* parameter)
This parameter specifies the number of sides per dice. When using 'XdY+Z' notation, this value is 'Y'. 

**Example**: if you are rolling three six-sided dice (3d6) to determine your stats, this value should be '6'.

### -Modifier (*add_val* parameter)
Each time you roll for an attribute (such as STR), this parameter specifies the amount that will be added or subtracted. When using 'XdY+Z' notation, this value is 'Z'. 

*Note*- this same modifier is applied to every attribute, up to the total 'Number of Attributes'. For instance, if you roll 6 attributes with a 'Modifier' of -1, the -1 will be applied to each attribute individually. This is not the same thing as racial modifiers.

**Default**: '0'. 

**Example**: if you are rolling three four-sided dice and adding six (3d4+6) to determine your stats, this value should be '6'. 

### -Dice to Keep Per Attribute (*keep_dice* parameter)
This value allows you to roll more dice than you need, keeping only the best ones. Keeping less dice than the total rolled dice will increase the average point buy equivalent, since the worst die rolls will be discarded. When using 'XdY+Z' notation, this value affects the number of 'X' that will be kept at the end. Note that it's impossible for the number of dice to keep to be greater than the number of dice that were rolled. 

**Default**: if using the Python API and the *keep_dice* is not specified, the default is to use the same as the 'number of dice per attribute'. If using the Bokeh server, the default is '3'. 

**Example**: if you are rolling four six-sided dice and keeping the best three dice, the 'number of dice per attribute' should be '4' and the 'dice to keep per attribute' should be '3'.

### -Number of Attributes (*num_attribute* parameter)

This value allows you to adjust the number of character attributes which will be rolled. Most commonly, the number of attributes will be 6 (STR, DEX, CON, INT, WIS, CHA); however, some DMs may wish the characters to have additional attributes, such as 'comeliness'. Furthermore, an rpg system may require more than 6 attributes to be rolled at character creation. 

**Default**: '6'. 

**Example**: The DM wishes the characters to have seven attributes (STR, DEX, CON, INT, WIS, CHA, COM), so the 'number of attributes' should be '7', and the 'attributes to keep' should be '7' as well.

### -Attributes to Keep (*keep_attribute* parameter)
This value allows you to roll more attributes than you need, keeping only the best ones. Keeping less attributes than the total generated attributes will increase the average point buy equivalent, since the worst attributes will be discarded. Note that it's impossible for the number of attributes to keep to be greater than the number of attributes that were rolled. 

**Default**: '6'. 

**Example**: if you are rolling ten attributes but only keeping the best six attributes, the 'number of attributes' should be '10', and the 'attributes to keep' should be '6'.

### -Rerolls (*reroll* parameter)
Each time you roll a die, you may wish to reroll values that are too low. Any die result that is less than or equal to this parameter will be rerolled. If this value is '0' (the default), no die will be rerolled at all. Values greater than 1 are also inclusive of lower values (i.e. '3' means reroll 1s, 2s, and 3s). Increasing this value will also increase the average point buy equivalent, because the worst die rolls will be rerolled. 

**Default**: '0'. 

**Examples**: if are rolling three six-sided dice but rerolling any 1s that come up, this value should be '1'. If you are rerolling any 1s or 2s, this value should be '2'.

### -Number of Arrays (*num_arrays* parameter)

CHANGELOG.rst
===============
 Change Log
===============

Version 0.11
-------------
- Initial build

Version 0.12
-------------
- Updated plot titles
- Unit tests
- Documentation
- New logo

Version 0.13
-------------
- Support for 4e

Version 0.14
-------------
- Added a Jupyter notebook simulator (pypbe-nb)
- Added a executable simulator (pypbe-sim)

Version 0.15
-------------
- Added more unit tests
- Configuration error-checking during initialization of PBE object
- Better commenting of the core module
- Added a Bokeh server simulator (pypbe-bk)

Version 0.16
-------------
- Added the ability to specify lower and upper limits for ability scores and point buy equivalents
- Updated pypbe-bk with the limits mentioned above
- Updated documentation

pypbe/core.py
# -*- coding: utf-8 -*-
"""
    pypbe.core
    --------------
    Intended for tabletop rpgs such as Pathfinder or D&D, this module contains
    code to run a Monte Carlo simulation to determine the Point Buy Equivalent
    (PBE) for a given type of ability score rolling method, such as "Roll 4d6
    and drop the lowest".

    :copyright: (c) 2017 Eric Strong.
    :license: Refer to LICENSE.txt for more information.
"""

import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from numpy.random import randint


class PBE:
    """
    Initializes a Monte Carlo simulation to determine the equivalent
    Point Buy of an ability score rolling method.
    """
    def __init__(self, num_dice, dice_type, add_val=0, num_attribute=6,
                 num_arrays=1, reroll=0, keep_dice=None, keep_attribute=6,
                 pbe_map='pf', custom_pbe_map=None, roll_low_limit=None,
                 roll_high_limit=None, pbe_low_limit=None,
                 pbe_high_limit=None):
        """
        Initializes a Monte Carlo simulation to determine the equivalent
        Point Buy of an ability score rolling method.

        :param num_dice: num_dice- The number of dice to roll (i.e. this is
            the "3" in "3d6")
        :param dice_type: The type of dice to roll (i.e. six-sided,
            eight-sided, etc. This is the "6" in "3d6")
        :param add_val: The value to add to the dice roll. (i.e.
            this is the "8" in "1d10+8")
        :param num_ability: The number of ability scores to generate (e.g. 6)
        :param num_arrays: The number of ability scores arrays that can be
            chosen from. For instance, 2 arrays might allow the player to
            choose between [12,10,6,11,15,17] and [6,9,12,18,15,10]
        :param reroll: Allow dice re-rolling, cumulatively. "0" is no re-rolls,
            "1" is re-rolling 1s, and "2" is re-rolling 1s and 2s, and so on.
        :param best_dice: If you want to roll more dice than you need
            and then take the best N results. E.g. "Roll 4d6 and drop the
            lowest roll" would require a "3" here.
        :param best_ability: If you want to roll more abilities than you need
            and then take the best N results. E.g. "Roll 3d6 seven times,
            and take the best six times" would require a "6" here.
        :param pbe_map: This determines how much each ability score will
            "cost" in the Point Buy system. You supply a string here, and
            the default is Pathfinder. You can (currently) select Pathfinder:
            'pf', D&D 3e: '3e', D&D 4e: '4e', or D&D 5e: '5e'
        :param custom_pbe_map: Supply a custom Point Buy mapping
        :param roll_low_limit: lowest allowable roll for an attribute
            (lower rolls will be discarded)
        :param roll_high_limit: highest allowable roll for an attribute
        :param pbe_low_limit: lowest allowable equivalent Point Buy
        :param pbe_high_limit: highest allowable equivalent Point Buy
        """
        # Save the user-supplied properties
        self.num_dice = int(num_dice)
        self.dice_type = int(dice_type)
        self.add_val = int(add_val)
        self.num_attribute = int(num_attribute)
        self.num_arrays = int(num_arrays)
        self.reroll = int(reroll)
        self.keep_attribute = int(keep_attribute)
        self.roll_low_limit = roll_low_limit
        self.roll_high_limit = roll_high_limit
        self.pbe_low_limit = pbe_low_limit
        self.pbe_high_limit = pbe_high_limit
        # If a custom PBE map is supplied, use it; otherwise, find the
        # PBE mapping based on the user-supplied string
        pbe_map = str(pbe_map).lower()
        if custom_pbe_map:
            if type(custom_pbe_map) == dict:
                self.pbe_map = custom_pbe_map
            else:
                raise TypeError("Custom PBE map must be a dictionary.")
        else:
            if pbe_map == 'pf' or pbe_map == '3e' or pbe_map == '3.5e' or \
                            pbe_map == '4e' or pbe_map == '5e':
                self.pbe_map = self._find_pb_mapping(pbe_map)
            else:
                raise ValueError("PBE map string unrecognized.")
        # If keep dice wasn't supplied, it's equal to the number of dice
        # (i.e. we're keeping "all" the dice)
        if keep_dice:
            self.keep_dice = keep_dice
        else:
            self.keep_dice = num_dice
        # Check supplied user input for errors
        self._error_check()

    # This function will perform some basic error-checking on the user
    # input. For instance, the amount of dice to keep can't be greater
    # than the amount of dice we are actually rolling.
    def _error_check(self):
        if self.keep_dice > self.num_dice:
            raise ValueError("Number of dice to keep cannot be greater " +
                             "than the number of dice.")
        if self.keep_attribute > self.num_attribute:
            raise ValueError("Number of attributes to keep cannot be greater" +
                             " than the number of attributes.")
        if self.reroll >= self.dice_type:
            raise ValueError("Number to re-roll must be less than dice type.")
        # If the lowest possible value is below the lowest defined point buy,
        # PyPBE can't calculate it. Same for the highest value being above the
        # highest defined point buy.
        low_pos_val = self.keep_dice * 1 + self.add_val
        high_pos_val = (self.dice_type * self.keep_dice) + self.add_val
        low_def_val = min(self.pbe_map.keys())
        high_def_val = max(self.pbe_map.keys())
        if low_pos_val < low_def_val:
            raise ValueError("The lowest possible value is " +
                             str(low_pos_val) + ". PBE is not defined for " +
                             "values less than " + str(low_def_val) +
                             ". Please increase the number of dice (or " +
                             "dice to keep) or the add value.")
        elif high_pos_val > high_def_val:
            raise ValueError("The highest possible value is " +
                             str(high_pos_val) + ". PBE is not defined for " +
                             "values greater than " + str(high_def_val) +
                             ". Please decrease the number of dice (or " +
                             "dice to keep) or the add value.")
        # Check that the upper and lower roll limits are not impossible
        if self.roll_low_limit is not None:
            if self.roll_low_limit >= high_pos_val:
                raise ValueError("The highest possible value is " +
                             str(high_pos_val) + ", but the roll_low_limit " +
                             "was selected as " + str(self.roll_low_limit) + 
                             ". Please decrease the roll_low_limit.")
        if self.roll_high_limit is not None: 
            if self.roll_high_limit <= low_pos_val:
                raise ValueError("The lowest possible value is " +
                             str(low_pos_val) + ", but the roll_high_limit " +
                             "was selected as " + str(self.roll_high_limit) + 
                             ". Please decrease the roll_high_limit.")

    # This function will find a point buy mapping (dictionary) based on the
    # user-supplied string
    @staticmethod
    def _find_pb_mapping(map_s):
        if map_s == "pf":  # default
            vmap = {3: -16, 4: -12, 5: -9, 6: -6, 7: -4, 8: -2, 9: -1, 10: 0,
                    11: 1, 12: 2, 13: 3, 14: 5, 15: 7, 16: 10, 17: 13, 18: 17}
        elif (map_s == "3e") | (map_s == "3.5e") | (map_s == "5e"):
            vmap = {3: -7, 4: -5, 5: -3, 6: -2, 7: -1, 8: 0, 9: 1, 10: 2,
                    11: 3, 12: 4, 13: 5, 14: 6, 15: 8, 16: 10, 17: 13, 18: 16}
        elif map_s == "4e":
            vmap = {3: -12, 4: -9, 5: -7, 6: -5, 7: -3, 8: -2, 9: -1, 10: 0,
                    11: 1, 12: 2, 13: 3, 14: 5, 15: 7, 16: 9, 17: 12, 18: 16}
        # I would rather this function fail noisily, even though this input
        # is checked during object initialization
        else:
            raise ValueError("PBE map string unrecognized. (2)")
        return vmap

    # This function will roll a single stat array, with a number of
    # rows equal to [num_hist] and columns equal to [num_ability]
    @staticmethod
    def _roll_array(num_hist, num_dice, dice_type, add_val, num_ability,
                    best_dice, reroll, roll_low_limit, roll_high_limit):
        abilities = []
        # Rolls the full array containing [num_ability] Ability Scores
        for _ in range(num_ability):
            # Generate a matrix of random integers [num_hist] by [num_dice]
            attr = randint(reroll + 1, dice_type + 1,
                           size=(num_hist, num_dice))
            # Sorts and takes the top [best_dice] values.
            if best_dice < num_dice:
                attr = np.sort(attr, axis=1)[:, -best_dice:]
            # Sum the rolls together, and add the [add_val]
            result = np.add(attr.sum(axis=1), add_val)
            abilities.append(result)
        stats = np.column_stack(abilities)       
        # Remove all rolls lower than the low_limit. WARNING- this code 
        # REMOVES an entire ability roll if ANY of its elements do not meet
        # the criteria, instead of rerolling it. This can mean that there 
        # may be a lot fewer rolls in the final array, so the number of 
        # histories should be increased to compensate for the removed elements.
        if roll_low_limit is not None:
            stats = stats[(stats>=roll_low_limit).all(axis=1)]
        # Remove all rolls higher than the high_limit. Same warning as above
        if roll_high_limit is not None:
            stats = stats[(stats<=roll_high_limit).all(axis=1)]
        return stats

    # This function will roll multiple stat arrays, with dimensions:
    # raw_array = (num_arrays,num_hist,best_dice)
    # pbe_array = (num_hist,num_arrays)
    @staticmethod
    def _roll_arrays(num_hist, num_dice, dice_type, add_val, num_ability,
                     best_dice, reroll, num_arrays, best_ability, vmap,
                     roll_low_limit, roll_high_limit, pbe_low_limit, 
                     pbe_high_limit):
        arrays = []
        arrays_pbe = []
        # Iterates over the number of arrays
        for _ in range(num_arrays):
            array_ = PBE._roll_array(num_hist, num_dice, dice_type, add_val,
                                     num_ability, best_dice, reroll, 
                                     roll_low_limit, roll_high_limit)
            # Sorts the array- we will need this to determine the distributions
            # of the top ability score, second ability score, and so on
            array_ = np.sort(array_)
            # Takes the top [best_ability] values
            if best_ability < num_ability:
                array_ = array_[:, -best_ability:]
            # Finds the total PBE of the array
            array_pbe = np.sum(np.vectorize(vmap.__getitem__)(array_), axis=1)
            # Add to the list of arrays
            arrays.append(array_)
            arrays_pbe.append(array_pbe)    
        raw_arrays = np.array(arrays)
        pbe_arrays = np.column_stack(arrays_pbe)
        # Remove all pbe lower than the low_limit. WARNING- this code 
        # REMOVES an entire ability roll if ANY of its elements do not meet
        # the criteria, instead of rerolling it. This can mean that there 
        # may be a lot fewer rolls in the final array, so the number of 
        # histories should be increased to compensate for the removed elements.
        if pbe_low_limit is not None:
            keep_ind = (pbe_arrays>=pbe_low_limit).all(axis=1)
            raw_arrays = raw_arrays[:, keep_ind]
            pbe_arrays = pbe_arrays[keep_ind]
        # Remove all rolls higher than the high_limit. Same warning as above
        if pbe_high_limit is not None:
            keep_ind = (pbe_arrays<=pbe_high_limit).all(axis=1)
            raw_arrays = raw_arrays[:, keep_ind]
            pbe_arrays = pbe_arrays[keep_ind]    
        return raw_arrays, pbe_arrays
		
    # This function will roll multiple stat arrays and choose the one
    # with the highest PBE
    @staticmethod
    def _find_best_pbe(num_hist, num_dice, dice_type, add_val, num_ability,
                       best_dice, reroll, num_arrays, best_ability, vmap,
                       roll_low_limit, roll_high_limit, pbe_low_limit,
                       pbe_high_limit):
        raw_arrays, pbe_arrays = PBE._roll_arrays(num_hist, num_dice,
            dice_type, add_val, num_ability, best_dice, reroll, num_arrays,
            best_ability, vmap, roll_low_limit, roll_high_limit, pbe_low_limit,
            pbe_high_limit)
        # Find the index of the array with the best PBE
        shp = raw_arrays.shape
        y, x = np.ogrid[0:shp[1], 0:shp[2]]
        z = np.argmax(pbe_arrays, axis=1)
        zr = np.reshape(z, (-1, 1))
        # Select the array and PBE with the largest PBE, using the index above
        best_array = raw_arrays[zr, y, x]
        best_pbe = np.choose(z, pbe_arrays.T)
        # Find the bin count of each value.
        bin_arr = [np.bincount(best_array[:, i]) for i in range(best_ability)]
        bin_arr = [bin_arr[i] / sum(bin_arr[i]) for i in range(best_ability)]
        vals, counts = np.unique(best_pbe, return_counts=True)
        counts = counts / sum(counts)
        bin_pbe = np.array([[val, count] for val, count in zip(vals, counts)])
        # Summary statistics
        mean_arr = [np.mean(best_array[:, i])
                    for i in range(best_ability)]
        mean_pbe = np.mean(best_pbe)
        a5_arr = [int(np.percentile(best_array[:, i], 5))
                  for i in range(best_ability)]
        a5_pbe = int(np.percentile(best_pbe, 5))
        a95_arr = [int(np.percentile(best_array[:, i], 95))
                   for i in range(best_ability)]
        a95_pbe = int(np.percentile(best_pbe, 95))
        std_arr = [np.std(best_array[:, i])
                   for i in range(best_ability)]
        std_pbe = np.std(best_pbe)
        # The results are bundled together
        arr_res = {"raw_bins": bin_arr, "means": mean_arr, "stds": std_arr,
                   "5percentile": a5_arr, "95percentile": a95_arr}
        pbe_res = {"raw_bins": bin_pbe, "means": mean_pbe, "stds": std_pbe,
                   "5percentile": a5_pbe, "95percentile": a95_pbe}
        return arr_res, pbe_res

    # This function will construct a title for the plots, based on the
    # user-supplied input
    @staticmethod
    def _construct_title(num_dice, dice_type, add_val, num_ability,
                         best_ability, best_dice, reroll, num_arrays):
        # Construct part 1: "Sum AdB+CkD"
        dice_str = "Sum {}d{}".format(num_dice, dice_type)
        if add_val > 0:
            dice_str += "+{}".format(str(add_val))
        if best_dice < num_dice:
            dice_str += "k{}".format(str(best_dice))
        # Construct part 2: ", EkF Attrs"
        ability_str = ", {}".format(str(num_ability))
        if best_ability < num_ability:
            ability_str += "k{}".format(str(best_ability))
        ability_str += " Attrs"
        # Construct part 3: ", G Arrays"
        array_str = ""
        if num_arrays > 1:
            array_str += ", {} Arrays".format(str(num_arrays))
        # Construct part 4: ", Reroll Hs"
        rerolls_str = ""
        if reroll > 0:
            rerolls_str = ", Reroll "
            for ii in range(0, reroll):
                rerolls_str += str(ii + 1) + "s/"
            rerolls_str = rerolls_str[:-1]
        # Concatenate all four parts, and return the results
        ret = "".join([dice_str, ability_str, array_str, rerolls_str])
        return ret

    # This function will build the text box for the raw array plot, since we
    # need a new line for each roll
    @staticmethod
    def _build_text(mean_arr, a5_arr, a95_arr):
        ret_str = ""
        for ii in range(len(mean_arr)):
            app = 'Roll {}: {:.2f} [{:.0f},{:.0f}]\n'.format(
                str(ii + 1), mean_arr[ii], a5_arr[ii], a95_arr[ii])
            ret_str += app
        ret_str = ret_str[:-1]
        return ret_str

    # This function will plot a histogram using the bins found
    # from the roll_mc() function. roll_mc() must be called first
    @staticmethod
    def _plot_hist(arr_res, pbe_res, title, figsize):
        # Divide into two subplots, stacked horizontally
        # Props is a dictionary for the text box with summary stats
        f, ((ax1), (ax2)) = plt.subplots(2, 1, figsize=figsize)
        props = dict(boxstyle='round', facecolor='wheat', alpha=0.5)

        # Plot each raw array, which is ordered from smallest to largest
        # The max val is global and needed to scale the legend appropriately
        maxval = 0
        for ii, roll in enumerate(arr_res["raw_bins"]):
            ax1.plot(roll, label="Roll {}".format(ii + 1))
            maxval = max(maxval, max(roll))
        ax1.axis([1, 18, 0, 1.25 * maxval])

        # Plot the text box for the raw arrays, which contains the
        # means, 5th percentile, and 95th percentile values
        str1 = PBE._build_text(arr_res["means"], arr_res["5percentile"],
                               arr_res["95percentile"])
        ax1.text(0.05, 0.75, str1, transform=ax1.transAxes, bbox=props,
                 fontsize=12, verticalalignment='top',
                 horizontalalignment='left')

        # Plot the Point Buy Equivalent distribution
        ax2.plot(pbe_res["raw_bins"][:, 0], pbe_res["raw_bins"][:, 1],
                 label="PBE")
        maxval_pbe = 1.05 * max(pbe_res["raw_bins"][:, 1])
        ax2.fill_between(pbe_res["raw_bins"][:, 0],
                         pbe_res["raw_bins"][:, 1])
        meanval = pbe_res["means"]
        ax2.plot([meanval, meanval], [0, maxval_pbe], color='k',
                 ls='--', label="Mean")
        ax2.set_ylim([0, maxval_pbe])
        # Plot the 90% line
        five_ind = np.where(pbe_res["raw_bins"][:, 0] == pbe_res["5percentile"])[0]
        five_prob = pbe_res["raw_bins"][five_ind, 1]
        ax2.plot([pbe_res["5percentile"], pbe_res["95percentile"]],
                 [five_prob, five_prob], color='darkslategray',
                 ls=':', label="90%")
        # Plot the text box for the PBE, which contains the mean, std,
        # and 90% values
        str2 = '$\mu$= {:.1f}\n$\sigma$= {:.2f}\n5%= {:.0f}\n95%= {:.0f}'.format(
            pbe_res["means"], pbe_res["stds"], pbe_res["5percentile"],
            pbe_res["95percentile"])
        ax2.text(0.85, 0.90, str2, transform=ax2.transAxes, fontsize=12,
                 verticalalignment='top', bbox=props)

        # Labeling
        plt.suptitle(title, fontsize=14)
        ax1.legend(loc='upper center', ncol=len(arr_res["raw_bins"]))
        ax1.set_ylabel('Raw Values Probability')
        ax2.legend(loc='upper left')
        ax2.set_ylabel('Point Buy Equivalent Probability')
        return f

    def roll_mc(self, num_hist=10**5):
        """
        Runs the initialized Monte Carlo simulation to determine the equivalent
        Point Buy of an ability score rolling method. NOTE- This method must
        be called BEFORE "plot_histogram".

        This method returns self, so you can stack the "plot_histogram" and
        "get_results" methods, like pbe1.roll_mc().plot_histogram()

        :param num_hist: Number of Monte Carlo histories to run. Suggest 10**5
        """
        # This is hard to read. I wanted to keep the underlying functions
        # static, so this is essentially just a wrapper around _find_best_pbe.
        self.arr_res, self.pbe_res = self._find_best_pbe(num_hist,
            self.num_dice, self.dice_type, self.add_val, self.num_attribute,
            self.keep_dice, self.reroll, self.num_arrays, self.keep_attribute,
            self.pbe_map, self.roll_low_limit, self.roll_high_limit,
            self.pbe_low_limit, self.pbe_high_limit)
        return self

    def plot_histogram(self, title_prefix="", title_override="",
                       figsize=(8, 6)):
        """
        Plots a histogram of the results after the Monte Carlo simulation is
        run. NOTE- This method must be called AFTER "roll_mc".

        :param title_prefix: If desired, prefix the title (such as "Alg 1")
        :param title_override: Override the title string entirely
        :param figsize: The size of the histogram plot
        :return: a seaborn figure of the histogram
        """
        # Check that roll_mc has been called
        if not self.arr_res:
            raise ValueError("Call roll_mc before plotting the histogram.")
        # Find a title using either the override or _construct_title method
        if title_override:
            title = title_override
        else:
            title = title_prefix + PBE._construct_title(self.num_dice,
                self.dice_type, self.add_val, self.num_attribute,
                self.keep_attribute, self.keep_dice, self.reroll,
                self.num_arrays)
        # Construct the histogram
        f = self._plot_hist(self.arr_res, self.pbe_res, title, figsize)
        return f

    def get_results(self, title_prefix="", title_override="", rnd_dig=2):
        """
        Constructs a summary of the results as an array, which might be
        useful for writing the results of multiple algorithms to a table.
        NOTE- This method must be called AFTER "roll_mc".

        :param title_prefix: If desired, prefix the title (such as "Alg 1 ")
        :param title_override: Override the title string entirely
        :param rnd_dig: the number of digits to round to
        :return: A tuple of the raw array results and PBE results, as:
                 [Description, Typical Array, Mean, Std, 5%, 95%]
        """
        # Check that roll_mc has been called
        if not self.arr_res:
            raise ValueError("Call roll_mc before getting results.")
        # Find a title using either the override or _construct_title method
        if title_override:
            title = title_override
        else:
            ctitle = PBE._construct_title(self.num_dice, self.dice_type,
                        self.add_val, self.num_attribute, self.keep_attribute,
                        self.keep_dice, self.reroll, self.num_arrays)
            title = title_prefix + ctitle
        # Find the typical array
        typ_arr = "; ".join([str(round(x, rnd_dig))
                             for x in self.arr_res["means"]])
        res_row = [title, typ_arr,
                   round(self.pbe_res["means"], rnd_dig),
                   round(self.pbe_res["stds"], rnd_dig),
                   round(self.pbe_res["5percentile"], rnd_dig),
                   round(self.pbe_res["95percentile"], rnd_dig)]
        return res_row
